crates/bevy_pbr/src/render/mesh.rs

```diff
@@ -155,10 +170,62 @@ pub fn extract_meshes(
     commands.insert_or_spawn_batch(not_caster_values);
 }
 
+#[derive(Component)]
+pub struct SkinnedMeshJoints {
+    pub buffer: BufferVec<Mat4>,
+}
+
+impl SkinnedMeshJoints {
+    #[inline]
+    pub fn build(
+        skin: &SkinnedMesh,
+        inverse_bindposes: &Assets<SkinnedMeshInverseBindposes>,
+        joints: &Query<&GlobalTransform>,
+        max_joint_count: usize,
+    ) -> Option<Self> {
+        let inverse_bindposes = inverse_bindposes.get(&skin.inverse_bindposes)?;
+        let bindposes = inverse_bindposes.iter();
+        let skin_joints = skin.joints.iter();
+        let mut buffer = BufferVec::new(BufferUsages::UNIFORM);
+        for (inverse_bindpose, joint) in bindposes.zip(skin_joints).take(max_joint_count) {
+            let joint_matrix = joints.get(*joint).ok()?.compute_matrix();
+            buffer.push(joint_matrix * *inverse_bindpose);
+        }
+        Some(Self { buffer })
+    }
+}
+
+pub fn extract_skinned_meshes(
+    mut commands: Commands,
+    mut previous_len: Local<usize>,
+    query: Query<(Entity, &ComputedVisibility, &SkinnedMesh)>,
+    inverse_bindposes: Res<Assets<SkinnedMeshInverseBindposes>>,
+    joint_query: Query<&GlobalTransform>,
+    wgpu_settings: Res<WgpuSettings>,
+) {
+    let mut values = Vec::with_capacity(*previous_len);
+    let max_joint_count =
+        (wgpu_settings.limits.max_uniform_buffer_binding_size as usize) / JOINT_SIZE;
+    for (entity, computed_visibility, skin) in query.iter() {
+        if !computed_visibility.is_visible {
+            continue;
+        }
+        // TODO: This can be expensive, can we move this to prepare?
+        if let Some(uniform) =
+            SkinnedMeshJoints::build(skin, &inverse_bindposes, &joint_query, max_joint_count)
```

Comment:
Large per-entity CPU and GPU re-allocations per-frame isn't something we should be doing. This should either be pooled, or we should move to something like DynamicUniformVec, which is suspect is possible (and would simplify a lot of things, like removing the need for per-entity bind groups).

Reply:
Dynamic offsets impose per heavy alignment requirements. My understanding was that the space is really needed so that wouldn’t fly. But I agree. I’m sure there’s a way to reuse the buffer rather than recreating it.

Reply:
In this case, the alignment isn't an issue. `128 * 64 / 256 = 32`, which means these would be perfectly aligned.

Reply:
Yup. As we discussed on Discord, it would seem that the entire uniform buffer can be larger than 16kB, but the part that is bound is limited to 16kB.

I came up with an evil space-saving scheme on Discord - we use a ‘managed’ BufferVec<Mat4> To provide the data for a uniform buffer with dynamic offsets. In the shader we use something like array<mat4x4<f32>, 1024u> but we abuse that we only need to index into the first N of those 1024 for a given skin and we don’t use the rest. The shader doesn’t care what the bound data actually is. So we can write only the number of Mat4s that we need for the skin and no more. Then when writing the matrices for the next skin, we will pad with Mat4::ZERO up to the next 256-byte alignment, then push the Mat4s. Finally, when writing the buffer out, the buffer has to be at least the final dynamic offset + 1024 Mat4s large so that the final binding is within the buffer. Reuse the same BufferVec each frame, just rewrite its contents.

Reply:
Went with @superdump's idea here. This has an up to 192 byte overhead per skinned mesh and a potentially ~16KB overhead for the entire buffer, but this is both more memory efficient and uses only one bind group and buffer globally.